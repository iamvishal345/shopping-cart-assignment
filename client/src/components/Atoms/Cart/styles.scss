@import "assets/scss/_breakpoints.scss";
.cart-container {
  background-color: #eeeeee;
  padding: 12px 30px 12px 12px;
  display: flex;
  align-items: center;
  position: relative;
  .cart-icon {
    display: flex;
    align-items: center;
    cursor: pointer;
    svg {
      height: 24px;
      fill: #d10157;
    }
    > span {
      font-size: 14px;
      margin-left: 4px;
    }
  }
  .overlay {
    position: fixed;
    width: 100vw;
    height: 100vh;
    background-color: #000000;
    left: 0;
    top: 0;
    opacity: 0.6;
    z-index: 9;
    @include md {
      display: none;
    }
  }
  .cart-dropdown {
    position: absolute;
    top: 50px;
    background-color: #ffffff;
    width: 25rem;
    right: 0;
    z-index: 9;
    @include md {
      width: 100vw;
      height: calc(100vh - 80px);
      position: fixed;
      top: 80px;
    }
    @include xm {
      width: 100vw;
      height: calc(100vh - 48px);
      top: 48px;
    }
    .cart-header {
      padding: 1rem;
      background-color: #000000;
      color: #ffffff;
      display: flex;
      justify-content: space-between;
      @include md {
        background-color: #ffffff;
        color: #000000;
      }
      .cart-close-btn {
        border: none;
        background: none;
        color: #ffffff;
        @include md {
          display: none;
        }
      }
    }
    .cart-body {
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      height: calc(100vh - 150px);
      @include md {
        height: 78%;
      }
      @include xm {
        height: 80%;
      }
      .cart-empty {
        margin: auto;
        text-align: center;
        b {
          font-size: 18px;
        }
        p {
          margin: 0.5rem;
          font-size: 14px;
        }
      }
      .cart-with-products {
        background-color: #eeeeee;
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        height: 100%;
        .products-wrapper {
          height: 100%;
          max-height: 65vh;
          @include md {
            max-height: 78vh;
          }
          overflow-y: auto;
          .cart-product-card {
            display: flex;
            align-items: center;
            justify-content: space-between;
            padding: 0.5rem 1rem;
            background-color: #ffffff;
            margin-top: 8px;
            .details {
              flex: 1;
              margin: 0 8px;
              h5 {
                margin: 0;
                margin-bottom: 1rem;
              }
              .actions-container {
                display: flex;
                align-items: center;
                justify-content: space-between;
                width: 75%;
                max-width: 15rem;
                .btn {
                  margin: 0;
                  border-radius: 50%;
                  width: 20px;
                  padding: 0;
                  font-size: 19px;
                  @include md {
                    border-radius: 5px;
                  }
                }
              }
            }
            .final-price {
              align-self: flex-end;
              margin-bottom: 10px;
            }
          }
          .bottom-card {
            display: flex;
            padding: 8px;
            background-color: #ffffff;
            margin: 1rem 0.5rem;
            border-radius: 4px;
            p {
              margin: 6px 12px;
              font-size: 14px;
            }
          }
        }
        > p {
          padding: 4px 0;
          margin: 0;
          font-size: 14px;
          text-align: center;
          background-color: #ffffff;
        }
      }
      .btn {
        margin: 8px;
        width: auto;
        padding: 12px;
        border-radius: 4px;
        &.checkout {
          display: flex;
          justify-content: space-between;
          .arrow {
            margin-left: 8px;
          }
        }
      }
    }
  }
}
