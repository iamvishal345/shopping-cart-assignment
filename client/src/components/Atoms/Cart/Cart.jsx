import React, { useContext, useState } from "react";
import "./styles.scss";
import { ReactComponent as CartIcon } from "assets/cart.svg";
import { CartContext } from "context/CartContext";

export const Cart = () => {
  const productList = useContext(CartContext).products;
  const updateQuantity = useContext(CartContext).updateQuantity;
  const [showDropdown, setShowDropdown] = useState(false);
  const handleShowDropdown = (e) => {
    e.preventDefault();
    setShowDropdown(!showDropdown);
  };
  return (
    <div className="cart-container">
      <div
        tabIndex={0}
        role="button"
        className="cart-icon"
        aria-label={showDropdown ? "Hide Cart Dropdown" : "Show Cart Dropdown"}
        onClick={handleShowDropdown}
        onKeyUp={(e) => e.code === "Enter" && handleShowDropdown(e)}
      >
        <CartIcon />
        <span id="no-of-items">{productList.length} items</span>
      </div>
      {showDropdown && (
        <React.Fragment>
          <div className="overlay"></div>
          <div className="cart-dropdown">
            <div className="cart-header">
              <span>
                <b>My Cart</b>
                {productList.length > 0 && <>({productList.length} item)</>}
              </span>
              <button
                onClick={handleShowDropdown}
                className="cart-close-btn"
                value="close"
              >
                X
              </button>
            </div>
            <div className="cart-body">
              {productList.length === 0 && (
                <div className="cart-empty">
                  <b>No items in your cart</b>
                  <p>Your favorite items are just a click away</p>
                </div>
              )}

              {productList.length > 0 && (
                <div className="cart-with-products">
                  <div className="products-wrapper">
                    {
                      <>
                        {productList.map((product, i) => (
                          <div key={product.id} className="cart-product-card">
                            <div className="image">
                              <img
                                loading="lazy"
                                height="70"
                                src={product.imageURL}
                                alt={product.name}
                              />
                            </div>
                            <div className="details">
                              <h5>{product.name}</h5>
                              <div className="actions-container">
                                <button
                                  id={`dec-btn-${i}`}
                                  type="button"
                                  className="btn"
                                  onClick={() => updateQuantity(product, -1)}
                                >
                                  -
                                </button>
                                <b>{product.quantity}</b>
                                <button
                                  id={`inc-btn-${i}`}
                                  type="button"
                                  className="btn"
                                  onClick={() => updateQuantity(product, 1)}
                                >
                                  +
                                </button>
                                <span>X</span>
                                <span>Rs.{product.price}</span>
                              </div>
                            </div>
                            <div className="final-price">
                              Rs.{product.price * product.quantity}
                            </div>
                          </div>
                        ))}
                        <div className="bottom-card">
                          <img
                            height="30"
                            src="/static/images/lowest-price.png"
                            alt="Lowest Price Guaranteed"
                          />
                          <p>You won't find it cheaper anywhere</p>
                        </div>
                      </>
                    }
                  </div>
                  <p>Promo code can be applied on payment page</p>
                </div>
              )}

              <button
                id="checkout-button"
                onClick={handleShowDropdown}
                className={`btn ${productList.length > 0 ? "checkout" : ""}`}
              >
                {productList.length > 0 ? (
                  <>
                    <span>Proceed to Checkout</span>
                    <span>
                      Rs.
                      {productList.reduce(
                        (acc, curr) => acc + curr.quantity * curr.price,
                        0
                      )}
                      <span className="arrow">{">"}</span>
                    </span>
                  </>
                ) : (
                  "Start Shopping"
                )}
              </button>
            </div>
          </div>
        </React.Fragment>
      )}
    </div>
  );
};
