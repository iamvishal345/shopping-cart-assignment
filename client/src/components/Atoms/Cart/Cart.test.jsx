import React from "react";
import { shallow } from "enzyme";
import { Cart } from "./Cart";

const mockUpdateQuantity = jest.fn();
const products = [
  {
    name: "Fresho Kiwi - Green, 3 pcs",
    imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
    description:
      "Kiwis are oval shaped with a brownish outer skin. The flesh is bright green and juicy with tiny, edible black seeds.",
    price: 87,
    stock: 50,
    category: "5b6899953d1a866534f516e2",
    sku: "fnw-kiwi-3",
    id: "5b6c6a7f01a7c38429530883",
    quantity: 3,
  },
  {
    name: "Apple - Washington, Regular, 4 pcs",
    imageURL: "/static/images/products/fruit-n-veg/apple.jpg",
    description:
      "The bright red coloured and heart shaped Washington apples are crunchy, juicy and slightly sweet. Washington Apples are a natural source of fibre and are fat free.",
    price: 187,
    stock: 50,
    category: "5b6899953d1a866534f516e2",
    sku: "fnw-apple-4",
    id: "5b6c6aeb01a7c38429530884",
    quantity: 1,
  },
];
jest.mock("react", () => {
  const ActualReact = jest.requireActual("react");
  return {
    ...ActualReact,
    useContext: () => ({
      products: products,
      updateQuantity: mockUpdateQuantity,
    }),
  };
});
describe("Cart", () => {
  let wrapped;
  beforeEach(() => {
    wrapped = shallow(<Cart />);
  });

  it("should match the snapshot", () => {
    expect(wrapped).toMatchSnapshot();
  });

  it("should show number of items as length of products array", () => {
    expect(wrapped.find("#no-of-items").text()).toEqual("2 items");
  });

  it("should show cart over on click of cart button", () => {
    wrapped.find(".cart-icon").simulate("click", {});
    expect(wrapped.find("#no-of-items").text()).toEqual("2 items");
  });
});
