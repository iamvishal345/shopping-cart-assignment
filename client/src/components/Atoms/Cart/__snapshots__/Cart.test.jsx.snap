// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should match the snapshot 1`] = `
<div
  className="cart-container"
>
  <div
    aria-label="Show Cart Dropdown"
    className="cart-icon"
    onClick={[Function]}
    onKeyUp={[Function]}
    role="button"
    tabIndex={0}
  >
    <ForwardRef(SvgCart) />
    <span
      id="no-of-items"
    >
      2
       items
    </span>
  </div>
</div>
`;
